### PR TITLE
Manfred speck optimisation

### PR DESCRIPTION
**Pull request by mj3cheun**
## About
- [ ] This is a new component
- [X] I am adding a feature to an existing component, or improving an existing feature
- [ ] I am closing an issue

## Description of changes

- Fix memory leak when Speck component is repeatedly created and destroyed
- Debounce Speck dash prop.view updates
- Add new dependency `fast-memoize` and add memoization to Speck where beneficial
- Add support for molecule translation
- General performance improvements and bug fixes
- Rewrite how the Speck react component interacts with Speck.Interaction due to breaking change in Speck library (Note: this PR is not backwards compatible with older versions of Speck)

REQUIRES https://github.com/plotly/speck/pull/6 TO FUNCTION CORRECTLY

## Before merging
- [ ] I have gone through the [code review checklist](https://github.com/plotly/dash-component-boilerplate/blob/master/%7B%7Bcookiecutter.project_shortname%7D%7D/review_checklist.md)
- [ ] I have completed all of the [steps to take before merging](https://github.com/plotly/dash-bio/blob/master/.github/before_merging.md)
- [ ] I know how to [deploy to DDS](https://github.com/plotly/dash-bio/blob/master/.github/deploying_to_dds.md)

### CODE REVIEW

**Comment by shammamah** [comment]
@mj3cheun The PR you linked (plotly/speck#6) says that it needs this one merged in -- please let me know which one I should review first!

**Comment by mj3cheun** [comment]
ideally they should be review together, one requires the other and vice versa

File: src/lib/components/Speck.react.js [hunk 1]
```diff
@@ -36,122 +91,149 @@ export default class Speck extends Component {
...
+    }
+
+    propsReconcile() {
+        if(!equals(this.view, this.props.view)) {
+            this.props.setProps({view: viewClone(this.view)});
         }
+    }
 
-        const system = speckSystem.new();
+    propsReconcileSchedule() {
```
Comment (shammamah) [inline src/lib/components/Speck.react.js]: Could you add in a small comment describing what this does?

File: src/lib/components/Speck.react.js [hunk 2]
```diff
@@ -1,13 +1,63 @@
...
     speckPresetViews,
 } from 'speck';
 
+
+
+/**
+ * Define private functions and variables used in the Speck component.
+ **/
+
+const PROPS_RECONCILE_DEBOUNCE_TIME = 500;
```
Comment (shammamah) [inline src/lib/components/Speck.react.js]: Please add a short description of what this variable means!

File: src/lib/components/Speck.react.js [hunk 3]
```diff
@@ -36,122 +91,149 @@ export default class Speck extends Component {
...
+                this.propsReconcileSchedule();
+            },
+
+            getZoom: () => this.view.zoom,
+            setZoom: zoomVal => {
+                this.view = viewAssign(this.view, {zoom: zoomVal});
+                this.propsReconcileSchedule();
+            },
+
+            refreshView: () => (this.refreshView = true),
```
[the remark this answers is not in the document]
Comment (mj3cheun) [reply src/lib/components/Speck.react.js]: This function never existed before, the function above is passed to `speckInteractions`, which allows `speckInteractions` to schedule a `refreshView` whenever required. This is done by setting the `refreshView` class property to `true`, which is a flag that is read the next time the `loop` class function is called (which is ideally 60 times/sec).

**Comment by shammamah** [comment]
I will approve this when https://github.com/plotly/speck/pull/6 and #370 have been merged. Other than that, 💃

**Comment by shammamah** [comment]
@mj3cheun Also, please add an entry to the CHANGELOG for this, bump the version to `0.1.2-rc1`, and rebuild/republish the package on PyPI.

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,7 @@
 ### Fixed
 * Fixed issue with Speck not responding to preset changes.
 * Refactored Speck component and implemented general bug fixes.
+* Made Speck component compatible with refactored Speck library
```
Comment (shammamah) [inline CHANGELOG.md]:
```suggestion
* Made Speck component compatible with refactored Speck library.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -8,6 +8,7 @@
 ### Fixed
 * Fixed issue with Speck not responding to preset changes.
 * Refactored Speck component and implemented general bug fixes.
```
Comment (shammamah) [inline CHANGELOG.md]: This and the one below it might be better off under a `### Changed` heading!

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,13 +1,21 @@
 # Changelog
 
+## [0.1.2] - 2019-06-10
```
Comment (shammamah) [inline CHANGELOG.md]:
```suggestion
## [0.1.2] - 
```

There are still a few more things to add to this release!

**Comment by shammamah** [comment]
@mj3cheun Please make sure that you've regenerated files and then published to PyPI: 

```
npm install 
npm run build:all
sudo rm -r dist/* 
python setup.py sdist
twine upload dist/*
```

**Pull request merged**